mmdet/models/detectors/dab_detr.py

```diff
@@ -0,0 +1,144 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from typing import Dict, Tuple
+
+from mmengine.model import uniform_init
+from torch import Tensor, nn
+
+from mmdet.registry import MODELS
+from ..layers import SinePositionalEncoding
+from ..layers.transformer import (DabDetrTransformerDecoder,
+                                  DabDetrTransformerEncoder, inverse_sigmoid)
+from .detr import DETR
+
+
+@MODELS.register_module()
+class DABDETR(DETR):
+    r"""Implementation of `DAB-DETR:
+    Dynamic Anchor Boxes are Better Queries for DETR.
+
+    <https://arxiv.org/abs/2201.12329>`_.
+
+    Code is modified from the `official github repo
+    <https://github.com/IDEA-Research/DAB-DETR>`_.
+
+    Args:
+        random_refpoints_xy (bool): Whether to randomly initialize query
+            embeddings and not update them during training.
+            Defaults to False.
+        num_patterns (int): Inspired by Anchor-DETR. Defaults to 0.
+    """
+
+    def __init__(self,
+                 *args,
+                 random_refpoints_xy: bool = False,
+                 num_patterns: int = 0,
+                 **kwargs) -> None:
+        self.random_refpoints_xy = random_refpoints_xy
+        assert isinstance(num_patterns, int), \
+            f'num_patterns should be int but {num_patterns}.'
+        self.num_patterns = num_patterns
+
+        super().__init__(*args, **kwargs)
+
+    def _init_layers(self) -> None:
+        """Initialize layers except for backbone, neck and bbox_head."""
+        self.positional_encoding = SinePositionalEncoding(
+            **self.positional_encoding_cfg)
+        self.encoder = DabDetrTransformerEncoder(**self.encoder)
+        self.decoder = DabDetrTransformerDecoder(**self.decoder)
+        self.embed_dims = self.encoder.embed_dims
+        self.query_dim = self.decoder.query_dim
+        self.query_embedding = nn.Embedding(self.num_queries, self.query_dim)
+        if self.num_patterns > 0:
+            self.patterns = nn.Embedding(self.num_patterns, self.embed_dims)
+        self.nb_dec = self.decoder.num_layers
```

Comment:
seems we do not use this attribute. I suggest removing this attribute and just using `self.decode.num_layers` at the detector level.